src/hardware/serial_terminal.rs

```diff
@@ -0,0 +1,77 @@
+use super::UsbBus;
+
+static OUTPUT_BUFFER: bbqueue::BBBuffer<1024> = bbqueue::BBBuffer::new();
+
+pub struct SerialTerminal {
+    usb_device: usb_device::device::UsbDevice<'static, UsbBus>,
+    usb_serial: usbd_serial::SerialPort<'static, UsbBus>,
+    output: bbqueue::Consumer<'static, 1024>,
+}
+
+impl SerialTerminal {
+    pub fn new(
+        usb_device: usb_device::device::UsbDevice<'static, UsbBus>,
+        usb_serial: usbd_serial::SerialPort<'static, UsbBus>,
+    ) -> Self {
+        let (_producer, consumer) = OUTPUT_BUFFER.try_split().unwrap();
+
+        Self {
+            usb_device,
+            usb_serial,
+            output: consumer,
+        }
+    }
+
+    fn flush(&mut self) {
+        let read = match self.output.read() {
+            Ok(grant) => grant,
+            Err(bbqueue::Error::InsufficientSize) => return,
+            err => err.unwrap(),
+        };
+
+        match self.usb_serial.write(read.buf()) {
+            Ok(count) => read.release(count),
+            Err(usbd_serial::UsbError::WouldBlock) => read.release(0),
+            Err(_) => {
+                let len = read.buf().len();
+                read.release(len);
+            }
+        }
+    }
+
+    pub fn usb_is_suspended(&self) -> bool {
+        self.usb_device.state() == usb_device::device::UsbDeviceState::Suspend
+    }
+
+    pub fn process(&mut self) {
+        self.flush();
+
+        if !self.usb_device.poll(&mut [&mut self.usb_serial]) {
+            return;
+        }
+
+        let mut buffer = [0u8; 64];
+        match self.usb_serial.read(&mut buffer) {
+            Ok(count) => {
+                for &value in &buffer[..count] {
+                    // Currently, just echo it back.
+                    self.usb_serial.write(b"echo: ").ok();
+                    self.usb_serial.write(&[value]).ok();
+                    self.usb_serial.write(b"\n\r").ok();
+                }
+            }
+
+            Err(usbd_serial::UsbError::WouldBlock) => {}
+            Err(_) => {
+                // Flush the output buffer if USB is not connected.
```

Comment:
Then you should be able to match the error explicitly.
And call this `clear the output buffer`.

Reply:
It's not exactly clear to me which error would indicate a USB connection not being present from the list of errors: https://docs.rs/usbd-serial/latest/usbd_serial/enum.UsbError.html - it seems reasonable to me to just clear the output buffer unconditionally if we can't read from the buffer for any reason, as there's definitely some communication breakdown that likely is occurring. Not sure how we should handle the errors otherwise.